src/cargo/util/process_builder.rs

```diff
@@ -151,10 +152,13 @@ impl ProcessBuilder {
 
     /// Runs the process, waiting for completion, and mapping non-success exit codes to an error.
     pub fn exec(&self) -> CargoResult<()> {
-        let mut command = self.build_command();
-        let exit = command.status().chain_err(|| {
-            process_error(&format!("could not execute process {}", self), None, None)
-        })?;
+        let exit = match self.build_command().status() {
```

Comment:
Could this dance perhaps be folded into an internal `spawn()` method? That way this wouldn't have to be duplicated across `exec`, `exec_with_output`, and `exec_with_streaming`

Reply:
Additionally, the pattern of try-then-if-failed-try-again [didn't actually work out for rustc on Windows](https://github.com/rust-lang/rust/pull/47507) so I think we'll want to tweak the logic to roughly match what rustc does, unconditionally using a response file for "very likely to fail" requests on Windows.

Reply:
> Could this dance perhaps be folded into an internal spawn() method? That way this wouldn't have to be duplicated across exec, exec_with_output, and exec_with_streaming

Each call site uses different entry points / piping (status, output, and spawn) making this rather ugly.  https://github.com/rust-lang/cargo/pull/7762/commits/f3ff04b1c5d5671e27cea9f0cc146d5f57496ae9 folds what can be folded, but as you can see, it's actually a net _growth_ to code size.  I'm fine with keeping that change if you are though.  The new `spawn` and `output` fns only have a single call site each, but make the error chaining easier/cleaner.

If you can think of a better workaround for `Stdio` being unclonable than `spawn_with`'s callback, I'd love to hear it too!

> Additionally, the pattern of try-then-if-failed-try-again didn't actually work out for rustc on Windows so I think we'll want to tweak the logic to roughly match what rustc does, unconditionally using a response file for "very likely to fail" requests on Windows.

It's been working for cargo against my test repository, but I suppose the concern is a .cmd RUSTC_WRAPPER?

Reply:
> Could this dance perhaps be folded into an internal spawn() method? That way this wouldn't have to be duplicated across exec, exec_with_output, and exec_with_streaming

https://github.com/rust-lang/cargo/pull/7762/commits/f3ff04b1c5d5671e27cea9f0cc146d5f57496ae9 folds what can be folded, but it's actually a net *increase* to the amount of code.  Each of those fns uses slightly different entry points (status, output, spawn) and slightly different piping as a result.

The new status/output fns only have a single call site each, but at least they make the error chaining easier...?  I'd love to have a better alternative to spawn_with's callback, but Stdio doesn't seem to be clonable...

> Additionally, the pattern of try-then-if-failed-try-again didn't actually work out for rustc on Windows so I think we'll want to tweak the logic to roughly match what rustc does, unconditionally using a response file for "very likely to fail" requests on Windows.

It's been working in my testing, but I suppose the concern is a RUSTC_WRAPPER.cmd ?